### PR TITLE
Add encryption and decryption pipes to arg parser

### PR DESCRIPTION
**Pull request by holmanb**
* **What does this PR do?**
allow use of oauth2 script without manual modification, set executable bit on script

* **Validation**
I didn't see any existing tests (please correct me if I'm wrong). To minimally validate the behavior works as expected, I applied the following patch and tested some inputs:
```diff
diff --git a/contrib/oauth2/mutt_oauth2.py b/contrib/oauth2/mutt_oauth2.py
index 66b741ae3..1bc9f6687 100755
--- a/contrib/oauth2/mutt_oauth2.py
+++ b/contrib/oauth2/mutt_oauth2.py
@@ -103,6 +103,9 @@ args = ap.parse_args()
 
 ENCRYPTION_PIPE = args.encryption_pipe
 DECRYPTION_PIPE = args.decryption_pipe
+print(ENCRYPTION_PIPE)
+print(DECRYPTION_PIPE)
+import os; os._exit(0)
 
 token = {}
 path = Path(args.tokenfile)
```
note that `shlex` handles spaces in quoted arguments as one would expect:
```
> ./mutt_oauth2.py userid@myschool.edu.tokens --verbose --test --encryption-pipe 'gpg --encrypt --recipient "me plus you"' --decrypt 'gpg --decrypt "another arg"'
['gpg', '--encrypt', '--recipient', 'me plus you']
['gpg', '--decrypt', 'another arg']
```
an email won't have spaces, but since this may be an arbitrary command, shlex is good hygiene

default values:
```
> ./mutt_oauth2.py userid@myschool.edu.tokens --verbose --test
['gpg', '--encrypt', '--recipient', 'YOUR_GPG_IDENTITY']
['gpg', '--decrypt']
```
and updated help:
```
options:
  -h, --help            show this help message and exit
  -v, --verbose         increase verbosity
  -d, --debug           enable debug output
  -a, --authorize       manually authorize new tokens
  --authflow AUTHFLOW   authcode | localhostauthcode | devicecode
  -t, --test            test IMAP/POP/SMTP endpoints
  --decryption-pipe DECRYPTION_PIPE
                        decryption command (string) default: "gpg --decrypt"
  --encryption-pipe ENCRYPTION_PIPE
                        encryption command (string) default: "gpg --encrypt --recipient YOUR_GPG_IDENTITY"
```

* **Does this PR meet the acceptance criteria?** (This is just a reminder for you,
  this section can be removed if you fulfill it.)

   - Code follows the [style guide](https://neomutt.org/dev/code)
   
No mention of python that I see on this page, so I used best judgement, avoided additional dependencies, avoided new python3 features for broad compatibility, etc.

* **What are the relevant issue numbers?**

https://github.com/neomutt/neomutt/issues/3441

### CODE REVIEW

File: contrib/oauth2/mutt_oauth2.py [hunk 1]
```diff
@@ -92,8 +93,26 @@
...
+ap.add_argument('--microsoft-secret', type=str, default='', help='')
 args = ap.parse_args()
 
+ENCRYPTION_PIPE = args.encryption_pipe
+DECRYPTION_PIPE = args.decryption_pipe
+
+registrations['google']['client_id'] = args.google_id
+registrations['google']['client_secret'] = args.google_secret
+registrations['microsoft']['client_id'] = args.microsoft_id
+registrations['microsoft']['client_secret'] = args.microsoft_secret
```
Comment (brunnre8) [inline contrib/oauth2/mutt_oauth2.py:114]: Why do we need split flags? Client_id / client_secret should do no? Then just add a provider switch to pre populate the endpoints.

Also introducing options without a help seems somewhat bad form.

**Comment by holmanb** [comment]
@brunnre8 Thanks for the review. I think I've addressed your comments.

File: contrib/oauth2/mutt_oauth2.py [hunk 2]
```diff
@@ -92,8 +93,30 @@
...
+ap.add_argument('--provider', type=str, choices=registrations.keys(),
+                help='Specify provider to use.')
 args = ap.parse_args()
 
+ENCRYPTION_PIPE = args.encryption_pipe
+DECRYPTION_PIPE = args.decryption_pipe
+
+provider = registrations.get(args.provider)
+if provider:
+    if args.client_id:
```
[the remark this answers is not in the document]
Comment (flatcap) [reply contrib/oauth2/mutt_oauth2.py:115]: > If we don't care about backwards compatibility
> (i.e. users manually inserting their values into the script)

I think the answer's, 'No'.
Wherever we install the script, it's in a readonly location, so the user must be editing their own copy.

Your improvements mean that they can use it directly from the install.
**Future** updates would need to be backwards-compatible.

[the remark this answers is not in the document]
Comment (flatcap) [reply contrib/oauth2/mutt_oauth2.py:115]: > If we don't care about backwards compatibility
> (i.e. users manually inserting their values into the script)

I think the answer's, 'No'.
Wherever we install the script, it's in a readonly location, so the user must be editing their own copy.

Your improvements mean that they can use it directly from the install.
**Future** updates would need to be backwards-compatible.

**Review by flatcap — CHANGES_REQUESTED** [review]
Following your suggestion,
Please can you remove any references / conditions relating to a hard-coded GPG Id

**Comment by holmanb** [comment]
> Following your suggestion, Please can you remove any references / conditions relating to a hard-coded GPG Id

Done.

The latest update to this PR is mostly argparsing changes, since I think I previously got all the references in the docs.

```./contrib/oauth2/mutt_oauth2.py -h
usage: mutt_oauth2.py [-h] [-v] [-d] [-a] [--authflow AUTHFLOW] [-t] [--decryption-pipe DECRYPTION_PIPE]
                      [--encryption-pipe ENCRYPTION_PIPE] [--client-id CLIENT_ID] [--client-secret CLIENT_SECRET]
                      [--provider {google,microsoft}]
                      tokenfile

positional arguments:
  tokenfile             persistent token storage

options:
  -h, --help            show this help message and exit
  -v, --verbose         increase verbosity
  -d, --debug           enable debug output
  -a, --authorize       manually authorize new tokens
  --authflow AUTHFLOW   authcode | localhostauthcode | devicecode
  -t, --test            test IMAP/POP/SMTP endpoints
  --decryption-pipe DECRYPTION_PIPE
                        decryption command (string), reads from stdin and writes to stdout, default: "gpg --decrypt"
  --encryption-pipe ENCRYPTION_PIPE
                        encryption command (string), reads from stdin and writes to stdout, suggested: "gpg
                        --encrypt --recipient YOUR_GPG_IDENTITY"
  --client-id CLIENT_ID
                        Provider id from registration
  --client-secret CLIENT_SECRET
                        (optional) Provider secret from registration
  --provider {google,microsoft}
                        Specify provider to use.

This script obtains and prints a valid OAuth2 access token. State is maintained in an encrypted TOKENFILE. Run with
"--verbose --authorize --encryption-pipe 'foo@bar.org'" to get started or whenever all tokens have expired,
optionally with "--authflow" to override the default authorization flow. To truly start over from scratch, first
delete TOKENFILE. Use "--verbose --test" to test the IMAP/POP/SMTP endpoints.
```

**Comment by flatcap** [comment]
@brunnre8 Anything you'd like to add?
(your voice carries more weight than mine :-)

**Comment by tristancollins** [comment]
Sorry for the quick, and likely not that useful, message here. I've spent a few hours trying to get the version of the mutt_oauth2.py script from the neomutt repo working (forced to look at alternate authentication methods as my O365 tenant seems to have started disallowing app passwords...) with no joy. Only when I replaced the neomutt version with the version from the mutt repo have things started working. I hadn't noticed the two versions were different as the date in the header was identical.

Anyway, it seems there is a bug in the updated parsing that forms the ENCRYPTION_PIPE. I'll take a further look over the weekend.

**Comment by holmanb** [comment]
> when I replaced the neomutt version with the version from the mutt repo have things started working. I hadn't noticed the two versions were different

@tristancollins - What value are you setting for the `--encryption-pipe` arg?

**Pull request merged**